host/sgx/sgxquote.c

```diff
@@ -125,18 +163,213 @@ static bool _load_sgx_dcap_ql(void)
+        local_mapped = NULL;
+        local_uuid = NULL;
+        local_key_id = NULL;
+
+        OE_TRACE_INFO(
+            "key_id_count=%lu mapped=%lu\n",
+            att_key_id_num,
+            mapped_key_id_count);
+
+        result = OE_OK;
+    }
+
+done:
+    if (local_mapped)
+    {
+        oe_free(local_mapped);
+        local_mapped = NULL;
+    }
+    if (local_uuid)
+    {
+        oe_free(local_uuid);
+        local_uuid = NULL;
+    }
+    if (local_key_id)
+    {
+        oe_free(local_key_id);
+        local_key_id = NULL;
+    }
+    if (_quote_ex_library.load_result == OE_OK)
+        _quote_ex_library.load_result = result;
+
+    OE_TRACE_INFO(
+        "_load_quote_ex_library_once() result=%s\n",
+        oe_result_str(_quote_ex_library.load_result));
+
+    return;
+}
+
+// For choosing between the DCAP library and the quote-ex library for quote
+// generation: the quote-ex library will be used if it can be loaded
+// and initialized successfully, otherwise the DCAP library is used.
+
+static bool _use_quote_ex_library(void)
+{
+    static oe_once_type once = OE_H_ONCE_INITIALIZER;
+    oe_once(&once, _load_quote_ex_library_once);
+
+    return (_quote_ex_library.load_result == OE_OK);
+}
+
 oe_result_t oe_sgx_qe_get_target_info(
     const oe_uuid_t* format_id,
     const void* opt_params,
     size_t opt_params_size,
     uint8_t* target_info)
 {
-    oe_result_t result = OE_FAILURE;
+    oe_result_t result = OE_UNEXPECTED;
     quote3_error_t err = SGX_QL_ERROR_UNEXPECTED;
 
-    OE_UNUSED(format_id);
-    OE_UNUSED(opt_params);
-    OE_UNUSED(opt_params_size);
+    if (!format_id || !target_info)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    if (_use_quote_ex_library())
+    {
+        sgx_status_t status = SGX_ERROR_UNEXPECTED;
+        sgx_att_key_id_ext_t updated_key_id = {{0}};
+        size_t tmp_size = 0;
+        uint8_t* tmp_buffer = NULL;
+        sgx_target_info_t tmp_target_info;
+
+        sgx_att_key_id_ext_t* key_id = _format_id_to_key_id(format_id);
+        if (!key_id)
+            OE_RAISE(OE_UNSUPPORTED);
+
+        // Update key ID with input SP ID for EPID quoting
+        memcpy(&updated_key_id, key_id, sizeof(*key_id));
+        if (key_id->base.algorithm_id == SGX_QL_ALG_EPID_LINKABLE ||
+            key_id->base.algorithm_id == SGX_QL_ALG_EPID_UNLINKABLE)
+        {
+            if (opt_params && opt_params_size == sizeof(key_id->spid))
+                memcpy(updated_key_id.spid, opt_params, opt_params_size);
+        }
+
+        status = _quote_ex_library.sgx_init_quote_ex(
+            (sgx_att_key_id_t*)&updated_key_id,
+            &tmp_target_info,
+            &tmp_size,
+            NULL);
+
+        if (status != SGX_SUCCESS)
+            OE_RAISE_MSG(
+                OE_PLATFORM_ERROR,
+                "sgx_init_quote_ex(NULL) returned 0x%x\n",
+                status);
+
+        tmp_buffer = (uint8_t*)oe_malloc(tmp_size);
+        OE_TEST(tmp_buffer);
```

Comment:
I don't think we use OE_TEST in SDK code; only in test code. It would be better to use OE_CHECK instead.

Reply:
Totally agree! Will fix.